tests/integration_tests/functional/test_cpu_features.py

```diff
@@ -475,52 +451,28 @@ def test_cpu_wrmsr_snapshot(
     )
 
 
-def diff_msrs(before, after, column_to_drop):
+def check_msrs_are_equal(before_df, after_df):
     """
-    Calculates and formats a diff between two MSR tables.
+    Checks that reported MSRs and their values in the files are equal.
     """
-    # Drop irrelevant column
-    before_stripped = before.drop(column_to_drop, axis=1)
-    after_stripped = after.drop(column_to_drop, axis=1)
-
-    # Check that values in remaining columns are the same
-    all_equal = (before_stripped == after_stripped).all(axis=None)
-
-    # Arrange the diff as a side by side comparison of statuses
-    not_equal = (before_stripped != after_stripped).any(axis=1)
-    before_stripped.columns = ["MSR_ADDR", "Before"]
-    after_stripped.columns = ["MSR_ADDR", "After"]
-    diff = pd.merge(
-        before_stripped[not_equal],
-        after_stripped[not_equal],
-        on=["MSR_ADDR", "MSR_ADDR"],
-    ).to_string()
-
-    # Return the diff or an empty string
-    return diff if not all_equal else ""
-
 
-def check_msr_values_are_equal(before_msrs_fname, after_msrs_fname):
-    """
-    Checks that MSR statuses and values in the files are equal.
-    """
-    before = pd.read_csv(before_msrs_fname)
-    after = pd.read_csv(after_msrs_fname)
+    # We first want to see if the same set of MSRs are exposed in the microvm.
+    # Drop the VALUE columns and compare the 2 dataframes.
+    impl_diff = pd.concat(
+        [before_df.drop(columns="VALUE"), after_df.drop(columns="VALUE")],
+        keys=["before", "after"],
+    ).drop_duplicates(keep=False)
+    assert impl_diff.empty, f"\n {impl_diff}"
 
-    flt_before = before[~before["MSR_ADDR"].isin(MSR_EXCEPTION_LIST)]
-    flt_after = after[~after["MSR_ADDR"].isin(MSR_EXCEPTION_LIST)]
+    # Remove MSR that can change at runtime.
+    before_df = before_df[~before_df["MSR_ADDR"].isin(MSR_EXCEPTION_LIST)]
+    after_df = after_df[~after_df["MSR_ADDR"].isin(MSR_EXCEPTION_LIST)]
 
-    # Consider only values of MSRs which are present both before and after
-    flt = (flt_before["STATUS"] == "implemented") & (
-        flt_after["STATUS"] == "implemented"
-    )
-    impl_before = flt_before.loc[flt]
-    impl_after = flt_after.loc[flt]
-
-    status_diff = diff_msrs(before, after, column_to_drop="VALUE")
-    value_diff = diff_msrs(impl_before, impl_after, column_to_drop="STATUS")
-    assert not status_diff
-    assert not value_diff
+    # Compare values
+    val_diff = pd.concat(
+        [before_df, after_df], keys=["before", "after"]
+    ).drop_duplicates(keep=False)
+    assert val_diff.empty, f"\n {val_diff}"
```

Comment:
Not sure what pandas will print with that since it usually omits rows/columns if they don't fit the terminal, but maybe `val_diff.to_string()` is more complete?

Reply:
indeed. will change it soon. thanks!

Reply:
done